File: week1_baseline/dotnet/src/Boukensha.Core/Registry.cs
namespace Boukensha.Core;

public sealed class Registry(Context context)
{
    public ToolDefinition Tool(
        string name,
        string description,
        IReadOnlyDictionary<string, ToolParameter>? parameters,
        Func<IReadOnlyDictionary<string, object?>, Task<string>> handler)
    {
        var tool = new ToolDefinition(name, description, parameters ?? new Dictionary<string, ToolParameter>(), handler);
        context.RegisterTool(tool);
        return tool;
    }

    public bool Registered(string name) => context.Tools.ContainsKey(name);

    public async Task<string> DispatchAsync(string name, IReadOnlyDictionary<string, object?>? args = null)
    {
        if (!context.Tools.TryGetValue(name, out var tool))
        {
            throw new UnknownToolException($"unknown tool: {name}");
        }
        return await tool.Handler(args ?? new Dictionary<string, object?>());
    }
}

File: week1_baseline/python/src/boukensha/registry.py
from __future__ import annotations

from typing import Callable

from .errors import UnknownToolError
from .tool import Tool


class Registry:
    def __init__(self, context) -> None:
        self.context = context

    def tool(self, name: str, *, description: str, parameters: dict | None = None, block: Callable) -> Tool:
        tool = Tool(str(name), description, parameters or {}, block)
        self.context.register_tool(tool)
        return tool

    def registered(self, name: str) -> bool:
        return str(name) in self.context.tools

    def dispatch(self, name: str, args: dict | None = None):
        args = args or {}
        tool = self.context.tools.get(str(name))
        if tool is None:
            raise UnknownToolError(f"No tool registered as '{name}'")
        return tool.block(**args)
